src/java.desktop/share/classes/javax/swing/JPopupMenu.java

```diff
@@ -758,6 +764,62 @@ public void pack() {
         }
     }
 
+    // We rely on the X11 input grab mechanism, but for the Wayland session
```

Comment:
Whilst we need to do this, having all this code discussing Wayland here in JPopupMenu is really the wrong place.
This needs to be re-done in a way that is about the specific problem which is something like
SunToolkit.clickOutsideWillDismissPopup() 
and the word "Wayland" shouldn't appear anywhere here.

IIUC this fix is needed so that a number of tests that fail now because of this problem will be able to pass, so it would be good to get it in soon so I can get a clearer picture of what remaining test issues we have.

Reply:
Sure, moved to toolkit.

> IIUC this fix is needed so that a number of tests that fail now because of this problem will be able to pass, so it would be good to get it in soon so I can get a clearer picture of what remaining test issues we have.

This only affects manual tests, as automatic tests use the XTest API, which works correctly with X11 Input Grab.

Besides, this is a real use case that can be annoying. This fix doesn't fix the problem completely (we still can't hide the menu by clicking on the title bar of our own window or on non-focusable windows such as the desktop), but it definitely makes the user experience better.